src/charm.py

```diff
@@ -582,26 +597,98 @@ def _wp_is_installed(self):
         logger.debug("Check if WordPress is installed")
         return self._run_wp_cli(["wp", "core", "is-installed"]).return_code == 0
 
-    def _current_effective_db_info(self):
+    @property
+    def _config_database_config(self) -> types_.DatabaseConfig | None:
+        """Database configuration details from charm config.
+
+        Returns:
+            Database configuration required to establish database connection.
+            None if not exists.
+        """
+        if any(
+            self.model.config.get(key) for key in ["db_host", "db_name", "db_user", "db_password"]
+        ):
+            return types_.DatabaseConfig(
+                hostname=self.model.config.get("db_host"),
+                database=self.model.config.get("db_name"),
+                username=self.model.config.get("db_user"),
+                password=self.model.config.get("db_password"),
+            )
+        return None
+
+    @property
+    def _db_relation_database_config(self) -> types_.DatabaseConfig | None:
+        """Database configuration details from stored state.
+
+        Since the legacy db relation is per-unit basis, the relation data must be stored in charm
+        state.
+
+        Returns:
+            Database configuration required to establish database connection.
+            None if not exists.
+        """
+        relation = self.model.get_relation(self._LEGACY_DB_RELATION_NAME)
+        if not relation:
+            return None
+        return types_.DatabaseConfig(
+            hostname=self.state.relation_db_host,
+            database=self.state.relation_db_name,
+            username=self.state.relation_db_user,
+            password=self.state.relation_db_password,
+        )
+
+    def _parse_database_endpoints(self, endpoints: str | None) -> str | None:
+        """Retrieve a single database endpoint from database endpoints.
+
+        Args:
+            endpoints: Comma separated endpoints of format host:port
+
+        Returns:
+            Hostname of database running on port 3306. None if no endpoints are provided.
+            Note that WordPress will throw MySQL Error when supplying a port with the hostname
+            (i.e. host:port).
+        """
+        if not endpoints:
+            return None
+        urls = endpoints.split(",")
+        return urls[0].split(":")[0]
```

Comment:
The default mysql port for WordPress is 3306, should we throw an error here if the port provided by mysql charm is not listen on 3306?

Reply:
Indeed we should, thank you for the suggestion :)